File: components/credentials/src/main/java/org/cloudfoundry/credhub/entity/Credential.java
package org.cloudfoundry.credhub.entity;

import java.util.UUID;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.Table;

import org.apache.commons.codec.digest.DigestUtils;
import org.cloudfoundry.credhub.audit.AuditableCredential;
import org.hibernate.annotations.GenericGenerator;

import static org.cloudfoundry.credhub.constants.UuidConstants.UUID_BYTES;

@Entity
@Table(name = "credential")
public class Credential implements AuditableCredential {

  @Id
  @Column(length = UUID_BYTES, columnDefinition = "VARBINARY")
  @GeneratedValue(generator = "uuid2")
  @GenericGenerator(name = "uuid2", strategy = "uuid2")
  private UUID uuid;

  @Column(nullable = false)
  private String name;

  @Column(unique = true, nullable = false)
  private String checksum;

  // Needed for hibernate
  @SuppressWarnings("unused")
  Credential() {
    this(null);
  }

  public Credential(final String name) {
    super();
    setName(name);
  }

  @Override
  public UUID getUuid() {
    return uuid;
  }

  @Override
  public void setUuid(final UUID uuid) {
    this.uuid = uuid;
  }

  @Override
  public String getName() {
    return name;
  }

  @Override
  public void setName(final String name) {
    this.name = name;
    if (name != null) {
      setChecksum(DigestUtils.sha256Hex(name));
    }
  }

  public String getChecksum() {
    return checksum;
  }

  public void setChecksum(final String checksum) {
    this.checksum = checksum;
  }
}

File: components/credentials/src/main/kotlin/org/cloudfoundry/credhub/entity/Credential.kt
package org.cloudfoundry.credhub.entity

import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.Id
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import org.apache.commons.codec.digest.DigestUtils
import org.cloudfoundry.credhub.audit.AuditableCredential
import org.cloudfoundry.credhub.constants.UuidConstants.Companion.UUID_BYTES
import org.hibernate.annotations.GenericGenerator
import java.util.UUID

@Entity
@Table(name = "credential")
class Credential : AuditableCredential {
    @Id
    @Column(length = UUID_BYTES, columnDefinition = "VARBINARY")
    @GeneratedValue(generator = "uuid2")
    @GenericGenerator(name = "uuid2", strategy = "uuid2")
    override var uuid: UUID? = null

    @OneToMany(
        cascade = [CascadeType.REMOVE],
        mappedBy = "credential",
        fetch = FetchType.LAZY,
    )
    var credentialVersions: MutableList<CredentialVersionData<*>> =
        mutableListOf()

    @Column(nullable = false)
    override var name: String? = null
        set(name) {
            field = name
            if (name != null) {
                checksum = DigestUtils.sha256Hex(name)
            }
        }

    @Column(name = "name_lowercase", nullable = false, insertable = false)
    var nameLowercase: String? = null

    @Column(unique = true, nullable = false)
    var checksum: String? = null

    // Needed for hibernate
    internal constructor() : this(null) {}

    constructor(name: String?) {
        this.name = name
    }
}
